aspnet/fundamentals/app-state.rst

```diff
@@ -1,8 +1,252 @@
+
+.. image:: app-state/_static/no-session-established.png
+
+This default behavior is produced by the following middleware in ``Startup.cs``, which runs when requests are made that do not already have an established session:
+
+.. literalinclude:: app-state/sample/src/AppState/Startup.cs
+	:linenos:
+	:language: c#
+	:lines: 70-96
+	:dedent: 8
+	:emphasize-lines: 4,7,9-11,25-26
+
+``GetOrCreateEntries`` is a helper method that will retrieve a ``RequestEntryCollection`` instance from ``Session`` if it exists; otherwise, it creates the empty collection and returns that. The collection holds ``RequestEntry`` instances, which keep track of the different requests the user has made during the current session, and how many requests they've made for each path.
+
+.. literalinclude:: app-state/sample/src/AppState/Model/RequestEntry.cs
+	:linenos:
+	:language: c#
+	:lines: 8-13
+	:dedent: 4
+
+.. literalinclude:: app-state/sample/src/AppState/Model/RequestEntryCollection.cs
+	:linenos:
+	:language: c#
+	:lines: 7-
+	:dedent: 4
+
+.. note:: The types that are to be stored in session must be marked with ``[Serializable]``.
+
+Fetching the current interest of ``RequestEntryCollection`` is done via the ``GetOrCreateEntries`` helper method:
+
+.. literalinclude:: app-state/sample/src/AppState/Startup.cs
+	:linenos:
+	:language: c#
+	:lines: 99-117
+	:dedent: 8
+	:emphasize-lines: 4,7-11
+
+When the entry for the object exists in ``Session``, it is retrieved as a ``byte[]`` type, and then deserialized using a ``MemoryStream`` and a ``BinaryFormatter``, as shown above. If the object isn't in ``Session``, the method returns a new instance of the ``RequestEntryCollection``.
+
+In the browser, clicking the Establish session hyperlink makes a request to the path "/session", and returns this result:
+
+.. image:: app-state/_static/session-established.png
+
+Refreshing the page results in the count incrementing; returning to the root of the site (after making a few more requests) results in this display, summarizing all of the requests that were made during the current session:
+
+.. image:: app-state/_static/session-established-with-request-counts.png
+
+Establishing the session is done in the middleware that handles requests to "/session":
+
+.. literalinclude:: app-state/sample/src/AppState/Startup.cs
+	:linenos:
+	:language: c#
+	:lines: 53-69
+	:dedent: 12
+	:emphasize-lines: 2,6-9
+
+Requests to this path will get or create a ``RequestEntryCollection``, will add the current path to it, and then will store it in session using the helper method ``SaveEntries``, shown below:
+
+.. literalinclude:: app-state/sample/src/AppState/Startup.cs
+	:linenos:
+	:language: c#
+	:lines: 118-126
+	:dedent: 8
+	:emphasize-lines: 3,5-7
+
+``SaveEntries`` demonstrates how to serialize a custom object into a ``byte[]`` for storage in ``Session`` using a ``MemoryStream`` and a ``BinaryFormatter``.
```

Comment:
I'd like to not tell people to use the BinaryFormatter. The BinaryFormatter is problematic when the input data does not come from a trusted source. I believe it's also not supported on .NET Core. Can we show using JSON.NET or the DataContractSerializer instead?